DSpace/DSpace
# [DS-4522] Logical item filtering (with DOI implementation)

## References
* JIRA: [DS-4522 JIRA ticket](https://jira.lyrasis.org/browse/DS-4522) ticket
* Fixes #7855 
* Documentation: [Wiki draft documentation page](https://wiki.lyrasis.org/pages/viewpage.action?spaceKey=~kshepherd&title=DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7)

## Description
This feature allows selective filtering of items for DOI minting, as well as ways for administrators to override/skip those filters, and brings with it a complete generic logical filtering system for DSpace items (similar in concept to the XOAI XML filter definitions) that can be used by any other service or component of DSpace.
A more thorough explanation is at [the draft wiki page](https://wiki.lyrasis.org/pages/viewpage.action?spaceKey=~kshepherd&title=DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7)

## Instructions for Reviewers
List of changes in this PR:
* Addition of classes in org.dspace.content.logic package, representing Logical Statements used in filtering
* Refactor of DOI Providers to make use of optional filters
* New _item-filters.xml_ spring configuration for defining item filters, with some examples already included
* Unit tests for logical filters, and updated tests for DOI providers to test filtering of items in DOI events
* Test command-line client for trying out filters without requiring them in production configuration

### To test: ###
* Have a read a of the [doco](https://wiki.lyrasis.org/pages/viewpage.action?spaceKey=~kshepherd&title=DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7) and get a feeling for how the filter spring configuration works, the difference between filters, operators, conditions, etc.
* Take a look at `item-filters.xml` and look for some filters to test, or define your own
* Use the command-line client to test out the filter over some items you know will pass or fail:
`dsrun org.dspace.content.logic.TestLogicRunner -f FILTERNAME -i HANDLE`
* Enable DOI providers / connectors the usual way, and include the filter in `identifier-service.xml` (there is a commented-out example)
* Test things normal operation with the DOI consumer, CLI tool, and make sure the filters are applied (unless explictly skipped with the CLI tool)
* Test out the curation task `registerdoi`, which by default should **always** bypass any filtering. It can be configured to obey filters instead, and to distribute across collections. (see `config/modules/doi-curation.cfg`)

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is not small, at well over 1000 lines long including the new classes, and DOI provider refactors. I hope an exception can be made for this new feature.
- [x] My PR includes no 3rd party dependencies
- [x] My PR does not modify REST contract or Angular UI

## Credits
This PR was developed by The Library Code GmbH with the support of Technische Informationsbibliothek (TIB) and the Technische Universität Hamburg TUHH.
